app/features/worksheet/tools.py

```diff
@@ -0,0 +1,278 @@
+def read_text_file(file_path):
+    # Get the directory containing the script file
+    script_dir = os.path.dirname(os.path.abspath(__file__))
+
+    # Combine the script directory with the relative file path
+    absolute_file_path = os.path.join(script_dir, file_path)
+    
+    with open(absolute_file_path, 'r') as file:
+        return file.read()
+
+# base class
+class QuestionBase:
+    def __init__(self, model, embedding_model, topic, grade_level, prompt_template, parser, verbose):
+        self.model = model
+        self.embedding_model = embedding_model
+        self.similarity_threshold = 0.9  # Adjust threshold as needed
+        self.topic = topic
+        self.grade_level = grade_level
+        self.prompt_template = prompt_template
+        self.parser = parser
+        self.verbose = verbose
+        self.set_names()
+        self.compile()
+    
+    def set_names(self):
+        self.section_name = ''
+        self.main_field = 'question'
+    
+    def compile(self, ):
+        # create the chain
+        prompt = PromptTemplate(
+            template=self.prompt_template,
+            input_variables=["topic", "existing_questions", "grade_level"],
+            partial_variables={"format_instructions": self.parser.get_format_instructions()}
+        )
+        
+        self.chain = prompt | self.model | self.parser
+        
+        if self.verbose: logger.info(f"Chain for {self.section_name} compilation complete")
+    
+    def sub_validate(self):
+        pass
+
+    def validate_response(self) -> bool:
+        try:
+            # Use Pydantic model to validate the response
+            if self.sub_validate():
+                return True
+            else:
+                return False
+        except ValidationError as e:
+            if self.verbose:
+                logger.error(f"Validation error during {self.section_name} response validation: {e}")
+            return False
+        except TypeError as e:
+            if self.verbose:
+                logger.error(f"TypeError during {self.section_name} response validation: {e}")
+            return False
+
+    def not_unique(self):
+        question = self.response[self.main_field]
+        question_embedding = self.embedding_model.encode(question, convert_to_tensor=True)
+
+        for q in self.question_bank:
+            similarity = util.pytorch_cos_sim(question_embedding, q)
+            if similarity.item() > self.similarity_threshold:
+                return True
+
+        self.question_bank.append(question_embedding)
+        self.existing_questions += f'\n{question}' 
+        return False
+    
+    def create_questions(self, num_questions = 0) -> List:
+        attempts = 0
+        max_attempts = num_questions * 5  # Allow for more attempts to generate questions
+        self.question_bank = []
+        self.existing_questions = ''
+        generated_questions = []
+
+        while len(generated_questions) < num_questions and attempts < max_attempts:
```

Comment:
Ensure that this loop doesn't generate too many tokens as it can be considered as costful.